### ISSUE TITLE
[ENH] Extending Reconciler to include methods which use the sample covariance of base forecast errors

### ISSUE THREAD

**Issue by Ajdin2022** [sktime/sktime#4239]
@ciaran-g, I would like to use `mint_cov` and `wls_var` methods to reconcile existing base forecasts in a hierachical setup. `Reconciler` currently only allows the methods `bu`, `ols`, `wls_str`, and `td_fcst`, none of which are based on the sample covariance of the base forecast errors (since these errors are not available to `Reconciler`). 

A solution to this could be to add a second argument to `Reconciler`'s `fit_transform` method - for the errors of the base forecasts. Namely, `Reconciler().fit_transform(preds, errors)`. Then the base forecasts could be reconciled using any of the methods that are available to `ReconcilerForecaster`. 

This would give freedom to developers to use any custom base forecasts in their hierachical setups, without having to give up on the best methods such as `mint_cov` and `wls_var`. Thanks!

**Comment by fkiraly** [comment]
hm, we can't change the signature of `fit_transform` because that is fixed - one of the benefits of `sktime` is that all estimators of a certain type have the same method signature, so you can plug & play.

If I understand correctly, @Ajdin2022, you want sth where you can input known forecast errors or historical actuals, rather than a forecastaer like in `ReconcilerForecaster` that still produces those?

Solutions I can think of:

* factor out a utility method from `ReconcilerForecaster` or add one
* make a variant of `mint_cov` which uses the series to be transformed for computation of the covariance - this is not the same as the `mint_cov` algorithm so would have to be explained in the docstring
* make an oracle forecast, e.g., `ForecastKnownValues` and plug that into a `ReconcilerForecaster`, and in turn plug the known values into the `ForecastKnownValues`

Would be interested in your opinion, @ciaran-g

**Comment by Ajdin2022** [comment]
@fkiraly, that is correct. I am not sure though that all the options you listed above solve it, because as input I would have to provide **2 things**, either

- base forecasts + base forecast errors, or 
- base forecasts + historical values

IMO, only then I can keep using my custom base forecasters, as well as do reconciliation based on the sample covariance of the base forecast errors, i.e., `mint_cov` method.

**Comment by fkiraly** [comment]
> IMO, only then I can keep using my custom base forecasters

Do you have some code that makes these forecasts?
Then a simple way could be, implement your forecasting logic as an `sktime` estimator, it should be easy to do!
https://www.sktime.net/en/stable/developer_guide/add_estimators.html
You can keep your custom forecaster local on your computer and private, you don't need to upload it to `sktime`, and still you can combine it with all the pipelines, aggregation, reconciliation logic in `sktime`.

That is, you fill in the extension template for the forecaster, and then run

```python
custom_forecaster = AjdinForecaster()
my_reconciler = ReconcilerForecaster(custom_forecaster method="mint_shrink")

my_reconciler.fit(y_historical, fh=my_fh)
my_reconciler.predict()
```

> I am not sure though that all the options you listed above solve it

"factor out" solves it because then you would simply run `factored_out_utility(y_historical, y_pred)` or similar

"oracle forecast" solves it, because then you would do:

```python
my_oracle = ForecastKnownValues(y_pred)
my_reconciler = ReconcilerForecaster(my_oracle, method="mint_shrink")

my_reconciler.fit(y_historical, fh=y_pred.index)
my_reconciler.predict()
```

**Comment by Ajdin2022** [comment]
Now I see how you would implement the oracle trick, @fkiraly, that's very clever! Factoring out solves it too, as you pass it both preds and the historical values. 

Thanks for the suggestion to add my own estimator. This would lead to complications on my end though - I have 3 different pipelines/algorithms creating forecasts for A, B, and C, which need to be reconciled at the end. So "oracle forecast" solves it in a much leaner way for me.

**Comment by fkiraly** [comment]
ok, I'll try to implement `ForecastKnownValues` then, maybe @ciaran-g also has a good idea.

> I have 3 different pipelines/algorithms creating forecasts for A, B, and C

You could use the `HierarchyEnsembleForecaster` for that, recently added by @VyomkeshVyas.
This allows you to use different forecasters for different levels or nodes at a hierarchy.

**Comment by Ajdin2022** [comment]
Adding `ForecastKnownValues` would be highly appreciated, @fkiraly! This is a great solution.

Thanks for the hint regarding `HierarchyEnsembleForecaster`, I didn't know about it. However as the models are already being trained in 3 different pipelines, I will keep `HierarchyEnsembleForecaster` in my mind for the next time!

**Comment by fkiraly** [comment]
> However as the models are already being trained in 3 different pipelines

I see. One maybe useful comment is (for next time), if you are working in deployment, it is worth keeping composability and interface homogeneity in mind also on the deployment/container layer. Doing too much of model specification manually may cause "proliferating pipeline design" that is very model specific and hard to maintain, engineering-wise; you may like to refactor-simplify to make the deployment architecture more flexible to model choice. Funny side note, I won't go too much into detail about technical situations at my current job, but if anyone from work reads this, they will also know what I mean :-)

**Comment by fkiraly** [comment]
draft for `ForecastKnownValues` here, feedback and testing appreciated:
https://github.com/sktime/sktime/pull/4243

**Comment by Ajdin2022** [comment]
Nice list of use-cases for this feature, @fkiraly! Thanks for the implementation. I will test it asap.

**Comment by Ajdin2022** [comment]
> > However as the models are already being trained in 3 different pipelines
> 
> I see. One maybe useful comment is (for next time), if you are working in deployment, it is worth keeping composability and interface homogeneity in mind also on the deployment/container layer. Doing too much of model specification manually may cause "proliferating pipeline design" that is very model specific and hard to maintain, engineering-wise; you may like to refactor-simplify to make the deployment architecture more flexible to model choice. Funny side note, I won't go too much into detail about technical situations at my current job, but if anyone from work reads this, they will also know what I mean :-)

@fkiraly, there are so many levels before reaching this one :D

**Comment by ciaran-g** [comment]
Hi both, as you have mentioned this is the use-case for the recently added `HierarchyEnsembleForecaster` (:

> You could use the `HierarchyEnsembleForecaster` for that, recently added by @VyomkeshVyas. This allows you to use different forecasters for different levels or nodes at a hierarchy.

I'm unclear how the oracle forecaster would solve the problem to be honest, since the `mint_cov` estimation relies on the residuals. Unless there's a way to do in-sample predictions or something obvious I'm missing? (quite possible lol)

**Comment by fkiraly** [comment]
> I'm unclear how the oracle forecaster would solve the problem to be honest

From above, this is how we would use it - the `ReconcilerForecaster` computes the residuals by calling `my_oracle`, which in turn plays back the known values from `y_pred`, as forecasts:

```python
my_oracle = ForecastKnownValues(y_pred)
my_reconciler = ReconcilerForecaster(my_oracle, method="mint_shrink")

my_reconciler.fit(y_historical, fh=y_pred.index)
my_reconciler.predict()
```

**Comment by fkiraly** [comment]
any luck with the PR prototype, @Ajdin2022?

**Comment by Ajdin2022** [comment]
Sorry for the delay, @fkiraly. I got stuck a bit. Namely, do I have to refit the `ReconcilerForecaster` (and relearn the `cov` matrix) after every point I add to the dataset? I plan to retrain my base forecasters very rarely, maybe once a year, so I would like to use the same cov matrix between these retrainings. (FYI, my data is very anomalous, and I ended up using boosting machines as base forecasters.)

**Comment by fkiraly** [comment]
You would do that using the `update` function. If you are using `ForecastKnownValues` (the current prototype), this will refit the covariance matrix only. (feedback appreciated whether that´s what you want)

**Comment by Ajdin2022** [comment]
I understand. So, the benefit of this approach is that the cov matrix would be always up-to-date, but at the cost of having to always maintain/update it, compared to the case in which a factored out utility from `ReconcilerForecaster` calculates the initial cov matrix once, and then another factored out utility just does the matrix multiplication to produce the reconciled forecasts. I think I will sync with my team to make a decision. Any recommendations from your side? :)

Btw, if I decide to use the first approach then it will be as follows, correct?

```
y_known      = base preds until now + next 3 hours
y_historical = hist values until now

my_oracle = ForecastKnownValues(y_known)
my_reconciler = ReconcilerForecaster(my_oracle, method="mint_cov")

my_reconciler.fit(y_historical, fh=[1,2,3])
my_reconciler.predict()
```

**Comment by fkiraly** [comment]
> Any recommendations from your side? :)

You can choose whether you want to update the cov matrix or not, by using the `update_params` (boolean) arg in `update`, or by wrapping the entire estimator in `DontUpdate` or `UpdateEvery` (to update only once every now and then).

> Btw, if I decide to use the first approach then it will be as follows, correct?

Yes, looks right to me.

**Comment by Ajdin2022** [comment]
Thanks a lot for your kind help!

**Comment by Ajdin2022** [comment]
To give my final feedback on this issue: the new forecaster `ForecastKnownValues` solves my problem, that is, I can use it within `ReconcilerForecaster` (as described above) to reconcile any given set of base forecasts in a hierachical setup, using any of the reconciliation methods available in `ReconcilerForecaster`. 

The advantage of this approach is that more reconciliation methods are available compared to 'Reconciler'.
